esi_leap/common/keystone.py

```diff
@@ -0,0 +1,43 @@
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+from keystoneauth1 import loading as ks_loading
+from keystoneclient import client as keystone_client
+
+import esi_leap.conf
+
+
+CONF = esi_leap.conf.CONF
+_cached_keystone_client = None
+
+
+def get_keystone_client():
+    global _cached_keystone_client
+    if _cached_keystone_client is not None:
+        return _cached_keystone_client
+
+    auth_plugin = ks_loading.load_auth_from_conf_options(CONF, 'keystone')
+    sess = ks_loading.load_session_from_conf_options(CONF, 'keystone',
+                                                     auth=auth_plugin)
+    cli = keystone_client.Client(session=sess)
+    _cached_keystone_client = cli
+
+    return cli
+
+
+def get_parent_project_id_tree(project_id):
```

Comment:
What is a parent project and why do we need to collect the parent project tree here?

Reply:
The model here is this: there are two separate projects that represent two different organizations - A and B. Each organization has multiple projects in their domain. A and B are node owners who may have temporary access to a given node.

This code is intended to solve the use case where A wants to offer up a node they own, but only to its own subprojects. In order to do so, they specify the lessee_id to be themselves, and this code will hide that offer from any project other than projects in their own tree.

Reply:
is lessee_id the best name for this? It wasnt immediately obvious to me that that would be the field name.

Reply:
Like something that implies its a family of ids rather than just one specific person. Just a random nit I have